src/Components/Web/src/Forms/InputFile/RemoteBrowserFileStream.cs

```diff
@@ -15,118 +13,46 @@ internal class RemoteBrowserFileStream : BrowserFileStream
+            _openRedStreamCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
 
-            var pipe = new Pipe(new PipeOptions(pauseWriterThreshold: options.MaxBufferSize, resumeWriterThreshold: options.MaxBufferSize));
-            _pipeReader = pipe.Reader;
-            _fillBufferCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
-
-            _ = FillBuffer(pipe.Writer, _fillBufferCts.Token);
+            OpenReadStreamTask = OpenReadStreamAsync(options, _openRedStreamCts.Token);
         }
 
-        private async Task FillBuffer(PipeWriter writer, CancellationToken cancellationToken)
+        private async Task<Stream> OpenReadStreamAsync(RemoteBrowserFileStreamOptions options, CancellationToken cancellationToken)
         {
-            long offset = 0;
-
-            while (offset < File.Size)
-            {
-                var pipeBuffer = writer.GetMemory(_maxSegmentSize);
-                var segmentSize = (int)Math.Min(_maxSegmentSize, File.Size - offset);
-
-                try
-                {
-                    using var readSegmentCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
-                    readSegmentCts.CancelAfter(_segmentFetchTimeout);
-
-                    var bytes = await _jsRuntime.InvokeAsync<byte[]>(
-                        InputFileInterop.ReadFileData,
-                        readSegmentCts.Token,
-                        _inputFileElement,
-                        File.Id,
-                        offset,
-                        segmentSize);
-
-                    if (bytes is null || bytes.Length != segmentSize)
-                    {
-                        throw new InvalidOperationException(
-                            $"A segment with size {bytes?.Length ?? 0} bytes was received, but {segmentSize} bytes were expected.");
-                    }
+            var dataReference = await _jsRuntime.InvokeAsync<IJSDataReference>(
+                InputFileInterop.ReadFileData,
+                cancellationToken,
+                _inputFileElement,
+                File.Id);
 
-                    bytes.CopyTo(pipeBuffer);
-                    writer.Advance(segmentSize);
-                    offset += segmentSize;
-
-                    var result = await writer.FlushAsync(cancellationToken);
-
-                    if (result.IsCompleted)
-                    {
-                        break;
-                    }
-                }
-                catch (Exception e)
-                {
-                    await writer.CompleteAsync(e);
-                    return;
-                }
-            }
-
-            await writer.CompleteAsync();
+            return await dataReference.OpenReadStreamAsync(_maxAllowedSize, options.MaxBufferSize, cancellationToken);
         }
 
         protected override async ValueTask<int> CopyFileDataIntoBuffer(long sourceOffset, Memory<byte> destination, CancellationToken cancellationToken)
```

Comment:
I'm not sure why it's valid to ignore the `sourceOffset` parameter now. Should that parameter be removed? What was it used for before?

Reply:
Actually I see it was ignored before too. Was that a bug? It's weird anyway.

Reply:
There are two implementations of the `BrowserFileStream`, this `RemoteBrowserFileStream` and the WebAssembly [`SharedBrowserFileStream` which does make use of this `sourceOffset`](https://github.com/dotnet/aspnetcore/blob/fcab188c61bbced0f523292cfc9d4accc438122a/src/Components/Web/src/Forms/InputFile/SharedBrowserFileStream.cs#L36).

We can ignore the `sourceOffset` here as the `stream` we're copying from handles that for us as we read and fill the `destination` buffer.